gunicorn: convert to Python packaging format + add Python3 variant

commodo opened the pull request:
Maintainer: me, @jefferyto (for Python parts)
Compile tested: https://github.com/openwrt/openwrt/commit/d6a405280f0156a2dad7d9cfa96695d53ed87dab  x86
Run tested: https://github.com/openwrt/openwrt/commit/d6a405280f0156a2dad7d9cfa96695d53ed87dab  x86

-------------------------------------------

And splitting the gunicorn[3] binary/executable away from the
python[3]-gunicorn libraries. This was inspired from Debian packaging.

The gunicorn[3] binaries require the new `python[3]-pkg-resources`
libraries to run, which add ~1.1 MB on the [ram]disk when uncompressed.

For the Python2 variant, not doing any bytecode compilation as it fails
with:
```
  File "/usr/lib/python2.7/site-packages/gunicorn/workers/_gaiohttp.py", line 84
    yield from self.wsgi.close()
             ^
  SyntaxError: invalid syntax

```

People around the web say this error is harmless.
 https://stackoverflow.com/questions/25611140/syntax-error-installing-gunicorn
 https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=803170
 https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=803202

Also, some Python[3] parts/features were added to better support some of
the quirks/needs of this entire change.

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

jefferyto commented on lang/python/gunicorn/Makefile:
```diff
@@ -10,43 +10,115 @@ include $(TOPDIR)/rules.mk
+
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(BUILD_VARIANT)-$(PKG_NAME)-$(PKG_VERSION)
+
+PYTHON_DO_NOT_COMPILE:=1
```
This will cause target Python to compile the code on each run, which isn't great for performance.

Is there any way we can work around or ignore the bytecode compilation error?

commodo replied:
yes; but this is how things were happening before this change for the Python2 version of gunicorn;
so, nothing is worse with regard to this change;

and i also had some internal thoughts about how-to-do things;
and i am glad that i can share some of them; makes me feel less insane/undecided

there are several other options:
1. compile everything except for `/usr/lib/python2.7/site-packages/gunicorn/workers/_gaiohttp.py`
2. patch `_gaiohttp.py` so that it compiles for Python2 as well; may not be great for Python3 as it may depend on that behavior
3. drop `_gaiohttp.py` from the package completely, until we move this to Python3

there may be other options here, but I can't think of them atm;

but the idea is that:
* Python2 should be going away
* the change should not make things worse for gunicorn (than they were before)
* i suspect people use gunicorn as a standalone-external sever (so gunicorn3 would be faster since it's byte-compiled)
* there are 2 users of gunicorn atm: `asu` and the seafile suite;  admit there may be users of gunicorn just using it for something else; but since it was ok for them without byte-compiled stuff, gunicorn3 should be better for them as well;

jefferyto replied:
I think Debian drops the file from the Python 2 package. That’s probably the easiest way to handle it.

I believe `PyPackage/python-gunicorn/filespec` is processed before the bytecode compilation step, so removing `_gaiohttp.py` should be relatively easy.

I think we can also keep the “do not compile” functionality, in case another package needs it.

commodo replied:
> I think Debian drops the file from the Python 2 package. That’s probably the easiest way to handle it.

hmmm; good idea;
i'll do that then


> I think we can also keep the “do not compile” functionality, in case another package needs it.

hmm; i wouldn't leave dead-code in there;
i would drop it

thanks for the ideas

jefferyto commented on lang/python/python-package.mk:
```diff
@@ -138,6 +139,10 @@ define PyBuild/Compile/Default
 
+define PyShebang
+$(SED) "1"'!'"b;s,^#"'!'".*python.*,#"'!'"/usr/bin/python2," -i --follow-symlinks $(1)
+endef
```
Having shebang handling in multiple places (here and in python-package-install.sh, and also the Py3 version) isn't ideal... Would it be possible to centralize it in one place (either in the shell script or makefiles)?

commodo replied:
yes it would; and it would a pretty good idea actually;
thanks for pointing it out

jefferyto commented on lang/python/python/files/python-package-setuptools.mk:
```diff
@@ -12,7 +12,7 @@ $(call Package/python/Default)
+  DEPENDS:=+python +python-pkg-resources
 endef
 
 define PyPackage/python-setuptools/install
```
I think there is also a line in `define PyPackage/python-setuptools/install` that copies `pkg_resources`?

commodo replied:
oops; thanks for catching this;

commodo commented:
Changelog v1 -> v2:
* removed `pkg_resources` from python-setuptools;
* unified Py[3]Shebang a bit better; 

will do some testing now and come back

commodo commented:
re-tested;

commodo commented:
@jefferyto 
ping; any more thoughts?

commodo commented:
Changelog v2 -> v3:
* fixed the Py[3]Shebang commit; add `&&` to the shell, otherwise the shebang script would mask the failure of the script
* dropped _do_not_compile_ logic; not needed
* not packaging _gaiohttp.py via filespec mechanism [forgot about it]

commodo commented:
i'm actually pretty happy with this final version :)

many thanks @jefferyto
